### PR TITLE
Improve performance of grok pattern cycle detection

### PR DESCRIPTION
**Pull request by masseyke**
We perform cycle detection on the bank of grok patterns that a user provides for the GrokProcessor. But it currently has some pretty severe performance problems (and bugs). This fixes that.

### CODE REVIEW

File: libs/grok/src/main/java/org/elasticsearch/grok/PatternBank.java [hunk 1]
```diff
@@ -112,25 +121,22 @@ private static void innerForbidCircularReferences(Map<String, String> bank, Stri
                 end = bracketIndex;
             } else if (columnIndex != -1 && bracketIndex == -1) {
                 end = columnIndex;
-            } else if (bracketIndex != -1 && columnIndex != -1) {
+            } else if (bracketIndex != -1) {
```
Comment (joegallo) [inline libs/grok/src/main/java/org/elasticsearch/grok/PatternBank.java:124]: For future GitHub archaeologists: this change looks out of place but isn't, IntelliJ warns that `columnIndex != -1` is always true here, so it's safe to have removed this.

**Review by joegallo — CHANGES_REQUESTED** [review]
```
        {
            Map<String, String> bank = new HashMap<>();
            bank.put("NAME1", "!!!%{NAME2} %{NAME3}!!!");
            bank.put("NAME2", "!!!%{NAME4} %{NAME5}!!!");
            bank.put("NAME3", "!!!!!!");
            bank.put("NAME4", "!!!!!!");
            bank.put("NAME5", "!!!%{NAME1}!!!");
            PatternBank.forbidCircularReferences(bank);
        }
```

This correctly detects a cycle (and fails) but the cycle explanation it returns with the `->`s isn't actually correct.

**Comment by masseyke** [comment]
> ```
>         {
>             Map<String, String> bank = new HashMap<>();
>             bank.put("NAME1", "!!!%{NAME2} %{NAME3}!!!");
>             bank.put("NAME2", "!!!%{NAME4} %{NAME5}!!!");
>             bank.put("NAME3", "!!!!!!");
>             bank.put("NAME4", "!!!!!!");
>             bank.put("NAME5", "!!!%{NAME1}!!!");
>             PatternBank.forbidCircularReferences(bank);
>         }
> ```
> 
> This correctly detects a cycle (and fails) but the cycle explanation it returns with the `->`s isn't actually correct.

There is now a test for this (and it passes).

**Review by joegallo — APPROVED** [review]
🚀

**Pull request merged**